common/migration/read.go

```diff
@@ -0,0 +1,163 @@
+
+type resultPair[T any] struct {
+	res T
+	err error
+}
+
+func (c *readerImpl[T]) ReadAndReturnActive(
+	ctx context.Context,
+	constraints Constraints,
+	old func(ctx context.Context) (*T, error),
+	new func(ctx context.Context) (*T, error),
+) (*T, error) {
+
+	state := c.getReaderRolloutState(constraints)
+
+	var active func(ctx context.Context) (*T, error)
+	var background func(ctx context.Context) (*T, error)
+
+	switch state {
+	case ReaderRolloutCallOnlyOldAndReturnOld:
+		return old(ctx)
+	case ReaderRolloutCallBothAndReturnOld:
+		active = old
+		background = new
+	case ReaderRolloutCallBothAndReturnNew:
+		active = new
+		background = old
+	case ReaderRolloutCallOnlyNewAndReturnNew:
+		return new(ctx)
+	default:
+		return old(ctx)
+	}
+
+	wg := sync.WaitGroup{}
+	wg.Add(2)
+
+	// 0 is always active data, index 1 is background
+	// using a slice to avoid locking
+	results := make([]resultPair[*T], 2)
+
+	// run the shadow request in a goroutine, wait for the results and then gather them up
+	go func() {
+		defer func() {
+			log.CapturePanic(recover(), c.log, nil)
+			c.scope.IncCounter(metrics.ShadowerBackgroundPanicRecoverCount)
+		}()
+		defer wg.Done()
+
+		backgroundRes, backgroundErr := c.callBackgroundAndCompare(ctx, background)
+		results[1].res = backgroundRes
+		results[1].err = backgroundErr
+	}()
+
+	// Fire off a watcher which is waiting for both sets of results to come through, and then once done so
+	// it'll hand them to the comparison function for analysis
+	go func() {
+		if !c.shouldCompare(constraints) {
+			return
+		}
+		defer func() {
+			log.CapturePanic(recover(), c.log, nil)
+			c.scope.IncCounter(metrics.ShadowerBackgroundPanicRecoverCount)
+		}()
+
+		wg.Wait()
+		c.comparisonFn(c.log, c.scope, results[0].res, results[0].err, results[1].res, results[1].err)
+	}()
+
+	activeRes, activeErr := active(ctx)
+	results[0].res = activeRes
+	results[0].err = activeErr
+	wg.Done()
+	return activeRes, activeErr
+}
+
+// Fires off a request on a background goroutine and waits for it to complete OR when the hard deadline
+// in backgroundTimeout expires. The op() code is expected to be faulty, to possibly panic, take too long
+// or otherwise misbehave and so the hard deadline is intended as a last-ditch safety measure if the context
+// is not being obeyed
+func (c *readerImpl[T]) callBackgroundAndCompare(parentCtx context.Context, op func(ctx context.Context) (*T, error)) (*T, error) {
+
+	newRequestCtx, cancel := copyParentContextWhileBreakingParentCancellation(parentCtx)
+	defer cancel()
+
+	timeoutCtx, cancel := context.WithTimeout(context.Background(), c.backgroundTimeout)
+	defer cancel()
+
+	completion := make(chan resultPair[*T])
+	go func() {
+		defer func() {
+			log.CapturePanic(recover(), c.log, nil)
+			c.scope.IncCounter(metrics.ShadowerBackgroundPanicRecoverCount)
+		}()
+		defer close(completion)
+
+		res, err := op(newRequestCtx)
+		completion <- resultPair[*T]{
+			res: res,
+			err: err,
+		}
+	}()
```

Comment:
currently this'll mean a panic closes -> reads an empty result-pair -> this is not treated as an error.

probably want something like
```go
go func() {
  var err error
  var res T
  defer func() {
    completion <- resultPair{res, err} // guarantees it writes
    close(completion) // no longer needed
  }()
  defer func() {
    r := recover()
    log.CapturePanic(r, c.log, &err)
    // maybe nothing was recovered
    if r != nil { c.scope.... }
  }()
  
  res, err = op(ctx)
}()
```

Reply:
good call... fixing